internal/runner/jest.go

```diff
@@ -0,0 +1,184 @@
+	}
+
+	if j.TestFileExcludePattern == "" {
+		j.TestFileExcludePattern = "node_modules"
+	}
+
+	if j.RetryTestCommand == "" {
+		j.RetryTestCommand = "yarn test --testNamePattern '{{testNamePattern}}' --json --testLocationInResults --outputFile tmp/jest.json"
+	}
+
+	return Jest{j}
+}
+
+func (j Jest) Name() string {
+	return "Jest"
+}
+
+// GetFiles returns an array of file names using the discovery pattern.
+func (j Jest) GetFiles() ([]string, error) {
+	debug.Println("Discovering test files with include pattern:", j.TestFilePattern, "exclude pattern:", j.TestFileExcludePattern)
+	files, err := discoverTestFiles(j.TestFilePattern, j.TestFileExcludePattern)
+	debug.Println("Discovered", len(files), "files")
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(files) == 0 {
+		return nil, fmt.Errorf("no files found with pattern %q and exclude pattern %q", j.TestFilePattern, j.TestFileExcludePattern)
+	}
+
+	return files, nil
+}
+
+func (j Jest) Run(testCases []string, retry bool) (RunResult, error) {
+	var cmd *exec.Cmd
+	var err error
+
+	defer os.Remove("tmp/jest.json")
+
+	if !retry {
+		commandName, commandArgs, err := j.commandNameAndArgs(j.TestCommand, testCases)
+		if err != nil {
+			return RunResult{Status: RunStatusError}, fmt.Errorf("failed to build command: %w", err)
+		}
+
+		fmt.Printf("%s %s\n", commandName, strings.Join(commandArgs, " "))
+		cmd = exec.Command(commandName, commandArgs...)
+	} else {
+		commandName, commandArgs, err := j.retryCommandNameAndArgs(j.RetryTestCommand, testCases)
+		if err != nil {
+			return RunResult{Status: RunStatusError}, fmt.Errorf("failed to build command: %w", err)
+		}
+
+		fmt.Println(commandName)
+		fmt.Println(commandArgs)
+
+		cmd = exec.Command(commandName, commandArgs...)
+	}
+
+	err = runAndForwardSignal(cmd)
+
+	if err == nil {
+		return RunResult{Status: RunStatusPassed}, nil
+	}
+
+	if ProcessSignaledError := new(ProcessSignaledError); errors.As(err, &ProcessSignaledError) {
+		return RunResult{Status: RunStatusError}, err
+	}
+
+	if exitError := new(exec.ExitError); errors.As(err, &exitError) {
+		report, parseErr := j.ParseReport("tmp/jest.json")
+		if parseErr != nil {
+			fmt.Println("Buildkite Test Splitter: Failed to read Jest output, tests will not be retried.")
+			return RunResult{Status: RunStatusError}, err
+		}
+
+		if report.NumFailedTests > 0 {
+			var failedTests []string
+			for _, testResult := range report.TestResults {
+				for _, example := range testResult.AssertionResults {
+					if example.Status == "failed" {
+						failedTests = append(failedTests, example.Name)
+					}
+				}
+			}
+
+			return RunResult{Status: RunStatusFailed, FailedTests: failedTests}, nil
+		}
+	}
+
+	return RunResult{Status: RunStatusError}, err
+}
+
+type JestExample struct {
+	Name   string `json:"fullName"`
+	Status string `json:"status"`
+}
+
+type JestReport struct {
+	NumFailedTests int `json:"numFailedTests"`
```

Comment:
```suggestion
	NumFailedTests int
```
[nit] I think we can omit this because the json attribute name is the same with the struct field name. Or if we want to keep this, maybe add the json field name for `TestResults` and `AssertionResults` to be consistent?